### ISSUE TITLE
Setting alternative default redshift for custom source?

### ISSUE THREAD

**Issue by djperrefort** [sncosmo/sncosmo#249]
I am currently using a custom source with ‘sncosmo’. Unfortunately, the spectra that I’m using to define the source assume the flux is measured some distance away. This is not in line with the default redshift value assumed by SNCosmo for the new source. When instantiating a new `Model` the redshift is set to `0`. Is there a preferred method for setting a different default redshift?

**Comment by kbarbary** [comment]
TL;DR: There's no built-in way to tell `Model` about what redshift its component `Source` prefers. As a workaround, you could create a wrapper function that creates a new `Model` and immediately sets the redshift to a new value. Another option would be to create a subclass of `Model` that sets the redshift as part of its `__init__` method. 

It's worth considering what the redshift has to do with the distance and brightness. As a high-redshift supernova person, I found I had a tendency to think about luminosity distance and redshift as directly related. However, this breaks down at low-z (For example, what is the luminosity distance at z=0? ... to a nearby blue-shifted galaxy?). And...  it's not an exact relation even at high redshift due to peculiar velocities. It's best to think about distance and redshift separately: Redshift *just* refers to the effect on the wavelength of the light (which can come from peculiar velocity or "cosmological" redshift) and distance is only apparent through the effect on apparent magnitude. The relation between the two at high redshift through the equation for D_L(z) arises due to the particular universe we happen to live in. If we lived in a different universe, one could have a luminosity distance and redshift that follow a completely different relation. sncosmo assumes nothing about the relation between redshift and distance/brightness, and separates the redshift part into `Model` and the brightness part into `Source`. 

But what about *distance* and brightness? These are directly related. A natural design would be to have `Source` specifying the absolute magnitude of the object. After all, `Source` is supposed to describe the rest-frame behavior of the source, with no knowledge of where it is relative to the observer. `Model` would add *three* parameters, not two: `z`, `t0`, and `dist`: `Model` would then be responsible for not only redshifting the spectrum, but also dimming the source appropriately for the distance it is at. While in many ways this would be a cleaner design, the distance and the absolute magnitude would usually be completely degenerate, so this would be an inconvenient parameterization for fitting the parameters to data. So instead, the source brightness parameter (e.g., `x0` or `amplitude`) represents the apparent brightness (effectively a combination of absolute brightness and distance), not the absolute brightness, even though it's sort of a weird design.

Sorry if that was a bit of a tangent... I hope it was related to your original question, or at least helps someone who has been wondering about this design aspect!

**Comment by djperrefort** [comment]
Not a tangent at all. This was extremely helpful. Thank you @kbarbary !
